File: packages/python/src/autonomous_shopping_optimizer/permits.py
from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass, field
from typing import Literal

RESOURCE_FIELDS = ("time", "tokens", "api_calls", "api_cost")
ExecutionStatus = Literal["completed", "timeout", "truncated", "cancelled", "failed"]


@dataclass(frozen=True)
class ResourceVector:
    time: int = 0
    tokens: int = 0
    api_calls: int = 0
    api_cost: int = 0

    def __post_init__(self) -> None:
        for resource in RESOURCE_FIELDS:
            value = getattr(self, resource)
            if isinstance(value, bool) or not isinstance(value, int) or value < 0:
                raise ValueError(f"{resource} must be a non-negative integer")

    @classmethod
    def from_mapping(cls, values: Mapping[str, object]) -> ResourceVector:
        return cls(**{resource: values.get(resource, 0) for resource in RESOURCE_FIELDS})

    def as_dict(self) -> dict[str, int]:
        return {resource: getattr(self, resource) for resource in RESOURCE_FIELDS}

    def fits_within(self, other: ResourceVector) -> bool:
        return all(
            getattr(self, resource) <= getattr(other, resource)
            for resource in RESOURCE_FIELDS
        )

    def add(self, other: ResourceVector) -> ResourceVector:
        return ResourceVector(
            **{
                resource: getattr(self, resource) + getattr(other, resource)
                for resource in RESOURCE_FIELDS
            }
        )

    def subtract(self, other: ResourceVector) -> ResourceVector:
        if not other.fits_within(self):
            raise ValueError("resource subtraction would produce a negative value")
        return ResourceVector(
            **{
                resource: getattr(self, resource) - getattr(other, resource)
                for resource in RESOURCE_FIELDS
            }
        )


@dataclass(frozen=True)
class PermitReservation:
    reservation_id: int
    merchant_id: str
    permit: ResourceVector
    _ledger_token: object = field(repr=False, compare=False)


@dataclass(frozen=True)
class UsageObservation:
    usage: ResourceVector
    exact_resources: frozenset[str]
    status: ExecutionStatus

    def __post_init__(self) -> None:
        unknown = self.exact_resources.difference(RESOURCE_FIELDS)
        if unknown:
            raise ValueError(f"unknown exact resources: {sorted(unknown)}")

    @classmethod
    def exact(
        cls,
        usage: ResourceVector,
        *,
        status: ExecutionStatus = "completed",
    ) -> UsageObservation:
        return cls(usage, frozenset(RESOURCE_FIELDS), status)


@dataclass(frozen=True)
class ReconciliationResult:
    reservation_id: int
    merchant_id: str
    status: ExecutionStatus
    observed_usage: ResourceVector
    charged_usage: ResourceVector
    reclaimed: ResourceVector
    remaining_budget: ResourceVector


class PermitLedger:
    """Atomically reserve hard budgets and conservatively reconcile call usage."""

    def __init__(self, budget: ResourceVector | Mapping[str, object]) -> None:
        self._initial = (
            budget if isinstance(budget, ResourceVector) else ResourceVector.from_mapping(budget)
        )
        self._remaining = self._initial
        self._ledger_token = object()
        self._next_reservation_id = 1
        self._active: dict[int, PermitReservation] = {}
        self._charged = ResourceVector()

    @property
    def initial_budget(self) -> ResourceVector:
        return self._initial

    @property
    def remaining_budget(self) -> ResourceVector:
        return self._remaining

    @property
    def charged_usage(self) -> ResourceVector:
        return self._charged

    def reserve(
        self,
        merchant_id: str,
        permit: ResourceVector | Mapping[str, object],
    ) -> PermitReservation:
        if not isinstance(merchant_id, str) or not merchant_id:
            raise ValueError("merchant_id must be a non-empty string")
        parsed_permit = (
            permit if isinstance(permit, ResourceVector) else ResourceVector.from_mapping(permit)
        )
        if not parsed_permit.fits_within(self._remaining):
            raise ValueError("permit exceeds the remaining budget")

        reservation = PermitReservation(
            reservation_id=self._next_reservation_id,
            merchant_id=merchant_id,
            permit=parsed_permit,
            _ledger_token=self._ledger_token,
        )
        self._remaining = self._remaining.subtract(parsed_permit)
        self._active[reservation.reservation_id] = reservation
        self._next_reservation_id += 1
        return reservation

    def reconcile(
        self,
        reservation: PermitReservation,
        observation: UsageObservation,
    ) -> ReconciliationResult:
        if reservation._ledger_token is not self._ledger_token:
            raise ValueError("reservation belongs to a different permit ledger")
        active = self._active.get(reservation.reservation_id)
        if active is None or active is not reservation:
            raise ValueError("reservation is unknown or already reconciled")
        if not observation.usage.fits_within(reservation.permit):
            raise ValueError("observed usage exceeds the reserved permit")

        charged = ResourceVector(
            **{
                resource: (
                    getattr(observation.usage, resource)
                    if resource in observation.exact_resources
                    else getattr(reservation.permit, resource)
                )
                for resource in RESOURCE_FIELDS
            }
        )
        reclaimed = reservation.permit.subtract(charged)
        self._remaining = self._remaining.add(reclaimed)
        self._charged = self._charged.add(charged)
        del self._active[reservation.reservation_id]

        return ReconciliationResult(
            reservation_id=reservation.reservation_id,
            merchant_id=reservation.merchant_id,
            status=observation.status,
            observed_usage=observation.usage,
            charged_usage=charged,
            reclaimed=reclaimed,
            remaining_budget=self._remaining,
        )

File: packages/javascript/src/permits.js
/**
 * Enforceable resource accounting for a shopping agent's tool loop.
 *
 * Reserve the full permit before dispatching a call, then reconcile once the call
 * returns. Capacity known to be unused is reclaimed; capacity whose true usage is
 * unknown, because the call was cancelled or truncated, is charged at the full
 * permit. Overruns are therefore prevented rather than detected afterwards.
 *
 * Mirrors the canonical Python `PermitLedger`.
 */

const RESOURCE_FIELDS = ["time", "tokens", "api_calls", "api_cost"];
const EXECUTION_STATUSES = new Set([
  "completed",
  "timeout",
  "truncated",
  "cancelled",
  "failed",
]);

function nonNegativeInteger(value, name) {
  if (!Number.isInteger(value) || value < 0) {
    throw new TypeError(`${name} must be a non-negative integer`);
  }
  return value;
}

/** Normalise a resource mapping, defaulting absent resources to zero. */
export function resourceVector(values = {}) {
  const vector = {};
  for (const field of RESOURCE_FIELDS) {
    vector[field] = nonNegativeInteger(values?.[field] ?? 0, field);
  }
  return vector;
}

function addVectors(left, right) {
  const total = {};
  for (const field of RESOURCE_FIELDS) total[field] = left[field] + right[field];
  return total;
}

function subtractVectors(left, right) {
  const difference = {};
  for (const field of RESOURCE_FIELDS) {
    const value = left[field] - right[field];
    if (value < 0) throw new RangeError("resource subtraction would go negative");
    difference[field] = value;
  }
  return difference;
}

function fitsWithin(candidate, limit) {
  return RESOURCE_FIELDS.every((field) => candidate[field] <= limit[field]);
}

export class PermitLedger {
  #token = Symbol("permit-ledger");
  #initial;
  #remaining;
  #charged;
  #active = new Map();
  #nextId = 1;

  constructor(budget) {
    this.#initial = resourceVector(budget);
    this.#remaining = this.#initial;
    this.#charged = resourceVector();
  }

  get initialBudget() {
    return { ...this.#initial };
  }

  get remainingBudget() {
    return { ...this.#remaining };
  }

  get chargedUsage() {
    return { ...this.#charged };
  }

  /** Deduct the whole permit up front, or fail without changing the ledger. */
  reserve(merchantId, permit) {
    if (typeof merchantId !== "string" || merchantId.length === 0) {
      throw new TypeError("merchantId must be a non-empty string");
    }
    const parsed = resourceVector(permit);
    if (!fitsWithin(parsed, this.#remaining)) {
      throw new RangeError("permit exceeds the remaining budget");
    }
    const reservation = {
      reservationId: this.#nextId,
      merchantId,
      permit: parsed,
      ledgerToken: this.#token,
    };
    this.#remaining = subtractVectors(this.#remaining, parsed);
    this.#active.set(reservation.reservationId, reservation);
    this.#nextId += 1;
    return reservation;
  }

  /**
   * Settle a reservation. `exactResources` lists the resources whose usage is
   * known; everything else is charged at the permit.
   */
  reconcile(reservation, { usage, exactResources = RESOURCE_FIELDS, status = "completed" }) {
    if (!reservation || reservation.ledgerToken !== this.#token) {
      throw new RangeError("reservation belongs to a different permit ledger");
    }
    const active = this.#active.get(reservation.reservationId);
    if (active === undefined || active !== reservation) {
      throw new RangeError("reservation is unknown or already reconciled");
    }
    if (!EXECUTION_STATUSES.has(status)) {
      throw new RangeError(`unsupported execution status: ${status}`);
    }
    const unknown = [...exactResources].filter((field) => !RESOURCE_FIELDS.includes(field));
    if (unknown.length > 0) {
      throw new RangeError(`unknown exact resources: ${unknown.sort().join(", ")}`);
    }
    const observed = resourceVector(usage);
    if (!fitsWithin(observed, reservation.permit)) {
      throw new RangeError("observed usage exceeds the reserved permit");
    }

    const exact = new Set(exactResources);
    const charged = {};
    for (const field of RESOURCE_FIELDS) {
      charged[field] = exact.has(field) ? observed[field] : reservation.permit[field];
    }
    const reclaimed = subtractVectors(reservation.permit, charged);
    this.#remaining = addVectors(this.#remaining, reclaimed);
    this.#charged = addVectors(this.#charged, charged);
    this.#active.delete(reservation.reservationId);

    return {
      reservationId: reservation.reservationId,
      merchantId: reservation.merchantId,
      status,
      observedUsage: observed,
      chargedUsage: charged,
      reclaimed,
      remainingBudget: { ...this.#remaining },
    };
  }
}
